### PR TITLE
Clean up IO module and its documentation

### PR DESCRIPTION
**Pull request by leekillough**
`channel.seek()` did not declare its offsets `int(64)`.

The list of `iomode` options were condensed into one table instead of being repeated in two adjacent paragraphs.

The `about-io-style` and `about-io-styles` sections were confusingly similar in name. `about-io-style` was really about `stdin-stdout-stderr`, so it was renamed to that.

Various grammatical and formatting improvements.

`.. TODO:` sections were added, which are considered comments in reStructuredText because the command `TODO` is not understood and is ignored. There, I listed options I thought needed to be considered for addition, such as corresponding to the `fopen()` options `"a"`, `"a+"`, `"wx"` and `"w+x"`. This is just my way of taking down notes and does not require an immediate decision, since it's commented out of the code and docs.
 
I also thought maybe `file` should have `isatty()` or `stat()` methods, but since they would have to be done in `qio`, I left them out.

I made the `iokind` options be documented in the order that their `enum` values were declared.

I fixed a documentation typo:
```
* ``==`` for set equality
* ``1=`` for set inequality
```
I added `pragma "no doc"` to the `file.init()` method, since it isn't really supported now, and I didn't want to see the big `TODO` in the official docs.

Fixed `other_exponent_char` documentation to list the correct value of `p` (thanks to @cassella for catching this).

Marked some of the older functions which use `error=` as deprecated.

### CODE REVIEW

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -816,27 +830,27 @@ extern record iostyle { // aka qio_style_t
...
+     Default is ``.``. */
   var point_char:style_char_t = 0x2e; // .
   /* When reading or writing a numeric value in a text mode channel,
-     how is the exponent written? Default is 'e' */
+     how is the exponent written? Default is ``e``. */
   var exponent_char:style_char_t = 0x65; // e
   /* When reading or writing a numeric value in a text mode channel,
-     when base is > 10, how is the exponent written? Default is 'e' */
+     when base is > 10, how is the exponent written? Default is ``e``. */
   var other_exponent_char:style_char_t = 0x70; // p
```
Comment (cassella) [inline modules/standard/IO.chpl:840]: The code says the default is `p`.

Comment (leekillough) [reply modules/standard/IO.chpl:840]: Good catch; I'll examine it further.

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -313,26 +313,24 @@ input, standard output, and standard error.
...
 
 .. _about-io-error-handling:
 
 Error Handling
 --------------
 
 Most I/O routines throw a :class:`SysError.SystemError`, and can be handled
-appropriately with ``try`` and ``catch``. For legacy reasons most I/O routines
+appropriately with ``try`` and ``catch``. For legacy reasons, most I/O routines
 can also can accept an optional `error=` argument.
```
Comment (mppf) [inline modules/standard/IO.chpl:325]: I think we've removed the `error=` versions a this point so that sentence should also be removed. It would be nice to remove other comments referring to the `error= version`.

File: modules/standard/IO.chpl [hunk 3]
```diff
@@ -2387,7 +2405,7 @@ inline proc channel.commit() where this.locking == false {
                          that the file is not seekable, or that the
                          channel is marked.
  */
-proc channel.seek(start:int, end:int = max(int)) throws {
+proc channel.seek(start:int(64), end:int(64) = max(int(64))) throws {
```
Comment (mppf) [inline modules/standard/IO.chpl:2408]: I'm not so sure why this change is important because `int` == `int(64)`. Can you share a rationale?

File: modules/standard/IO.chpl [hunk 4]
```diff
@@ -2494,7 +2512,7 @@ proc channel._set_style(style:iostyle) {
 /*
 
    Return the locale on which an ongoing I/O was started with a channel.
-   This method will return nil unless it is called on a channel that is
+   This method will return ``nil`` unless it is called on a channel that is
```
Comment (mppf) [inline modules/standard/IO.chpl:2515]: It's not actually ``nil`` because locales are now records. Looks like it uses `nilLocale` now. However `nilLocale` is not documented.

**Review by mppf — APPROVED** [review]
Thanks for looking over and improving the IO module docs.

File: modules/standard/IO.chpl (hunk 2 above)
Comment (leekillough) [reply modules/standard/IO.chpl:325]: These procedures still take a `ref error`, even if they are `#pragma "no doc"` (all but the last one):

```
pragma "no doc"
proc channel._match_regexp_if_needed(cur:size_t, len:size_t, ref error:syserr,
    ref style:iostyle, r:unmanaged _channel_regexp_info)
{
}

pragma "no doc"
proc channel._format_reader(
    fmtStr:?fmtType, ref cur:size_t, len:size_t, ref error:syserr,
    ref conv:qio_conv_t, ref gotConv:bool, ref style:iostyle,
    ref r:unmanaged _channel_regexp_info?,
    isReadf:bool)
{
}

pragma "no doc"
proc channel._conv_helper(
    ref error:syserr,
    ref conv:qio_conv_t, ref gotConv:bool,
    ref j:int,
    argType: c_ptr(c_int),
    argTypeLen: int)
{
}

pragma "no doc"
proc channel._conv_sethandler(
    ref error:syserr,
    argtypei:c_int,
    ref style:iostyle,
    i:int, argi,
    isReadf:bool):bool throws
{
}

pragma "no doc"
proc channel._extractMatch(m:reMatch, ref arg:reMatch, ref error:syserr) {
  // If the argument is a match record, just return it.
  arg = m;
}

pragma "no doc"
proc channel._extractMatch(m:reMatch, ref arg:bytes, ref error:syserr) {
}

pragma "no doc"
proc channel._extractMatch(m:reMatch, ref arg:?t, ref error:syserr)
      where t != reMatch && t != bytes {
}

// documented in throws version
pragma "no doc"
proc channel.extractMatch(m:reMatch, ref arg, ref error:syserr) {
}

pragma "no doc"
proc channel._ch_handle_captures(matches:_ddata(qio_regexp_string_piece_t),
                                 nmatches:int,
                                 ref captures, ref error:syserr) {
}

pragma "no doc"
proc channel.search(re:regexp(?), ref error:syserr):reMatch
{
}

pragma "no doc"
proc channel.match(re:regexp(?), ref error:syserr):reMatch
{
}

proc channel.match(re:regexp(?), ref captures ...?k, ref error:syserr):reMatch
{
}
```

Comment (mppf) [reply modules/standard/IO.chpl:325]: Right, since these are almost all no-doc we don't need to describe them in the module docs.

The last one should get a deprecation warning if it doesn't already have one.

Comment (mppf) [reply modules/standard/IO.chpl:325]: What I really mean is - some of these are internal (the ones beginning with `_`) and those we can leave as -is. The others should get a deprecation warning.

File: modules/standard/IO.chpl (hunk 3 above)
Comment (leekillough) [reply modules/standard/IO.chpl:2408]: I thought `int` was platform-dependent 32-bit or 64-bit, while `int(8)`, `int(16)`, `int(32)` and `int(64)` were guaranteed in size.

Does `int == int(64)` simply because we only target x86-64 and 64-bit GPU processors now, the reasoning being that `int` should always be large enough for any an array size or pointer difference? That seems orthogonal to the file offset issue.

Since this is for file offsets, it doesn't matter whether we're on an ILP32 or LP64 ABI model -- the file offsets should always be 64-bit, even on a ILP32 ABI like x86.

Plus, `int(64)` is used everywhere else offsets are used in the declarations and generated documentation of `IO`, so leaving it out of `channel.seek()` stood out.

Comment (mppf) [reply modules/standard/IO.chpl:2408]: `int` does not depend on platform (of course it does in C). See https://chapel-lang.org/docs/master/language/spec/types.html#signed-and-unsigned-integral-types

> Plus, int(64) is used everywhere else offsets are used in the declarations and generated documentation of IO, so leaving it out of channel.seek() stood out.

This is an OK rationale

File: modules/standard/IO.chpl (hunk 2 above)
Comment (leekillough) [reply modules/standard/IO.chpl]: Some of them simply appear to be overloads with 0 captures and with >0 captures, with the 0-capture ones being `pragma "no doc"`. This is all to get around the limitation that varargs can't work with 0 arguments.

Some of them are still used by some of the other modules. For the scope of this PR, I don't want to have to deprecate and refactor the other modules' usage.

Please see commit https://github.com/chapel-lang/chapel/pull/17448/commits/8444714fad9bdd805d08be5e114068cedf6ddb53

File: modules/standard/IO.chpl (hunk 4 above)
Comment (leekillough) [reply modules/standard/IO.chpl]: I changed it to `nilLocale` even if that's not documented elsewhere, for accuracy.
